src/unix.rs

```diff
@@ -493,3 +578,103 @@ extern "C" fn sigusr1_handler(
 ) {
     // nothing to do
 }
+
+#[cfg(target_os = "linux")]
+fn cvt_ssize(t: libc::ssize_t) -> io::Result<libc::ssize_t> {
+    if t == -1 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(t)
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn non_blocking_read(fd: c_int, buf: &[u8]) -> io::Result<usize> {
+    static IS_NONBLOCKING_READ_UNSUPPORTED: AtomicBool = AtomicBool::new(false);
+
+    if IS_NONBLOCKING_READ_UNSUPPORTED.load(Ordering::Relaxed) {
+        return Err(io::ErrorKind::Unsupported.into());
+    }
+
+    match cvt_ssize(unsafe {
```

Comment:
BTW, this is interesting: https://man.archlinux.org/man/preadv2.2.en#BUGS

> Linux 5.9 and Linux 5.10 have a bug where preadv2() with the RWF_NOWAIT flag may return 0 even when not at end of file.

See also <https://lkml.org/lkml/2021/5/24/255>. I am not sure if we need to do something here, or just warn user this may happen.

Reply:
AIUI `RWF_NOWAIT` support for pipes was only introduced in https://github.com/torvalds/linux/commit/afed6271f5b0d78ca1a3739c1da4aa3629b26bba which is newer than the buggy versions.

Reply:
Good to know that. Thanks!